### PR TITLE
[BUG] Fix KeyError in NeuralForecastLSTM for hierarchical data prediction #7691

### LINKED ISSUE #7691: [BUG] Unexpected `KeyError` when predicting hierarchical data with NeuralForecastLSTM
**Describe the bug**
While trying to forecast hierarchical data with NeuralForecastLSTM, I obtain an unexpected KeyError. I'm wondering why the tests haven't detected this.

**To Reproduce**

```python
import pandas as pd
from sktime.transformations.hierarchical.aggregate import Aggregator
df = pd.read_csv("https://raw.githubusercontent.com/felipeangelimvieira/prophetverse/f61e948d2840f73688570561218d2cf810a256bc/src/prophetverse/datasets/raw/tourism.csv")
df["Quarter"] = pd.to_datetime(df["Quarter"]).dt.to_period("Q")
y = df.set_index(["State", "Region", "Purpose", "Quarter"])[["Trips"]]
y = y.groupby(level=["Purpose", "State", "Quarter"]).sum()
y = Aggregator().fit_transform(y)
y

from sktime.forecasting.neuralforecast import NeuralForecastLSTM

lstm = NeuralForecastLSTM(
    freq="Q",
    trainer_kwargs={
        "accelerator" :"cpu"
    }
)

lstm.fit(y, fh=fh)
# Raises keyerror
y_pred = lstm.predict(fh=fh)
```

**Expected behavior**
Forecasts should be returned

**Additional context**
It seems that the error does not happen with single timeseries. If we replace `y` with `y.loc[("__total", "__total")]`, it works as expected

**Versions**
<details>
System:
    python: 3.11.11 (main, Dec 26 2024, 12:31:23) [Clang 16.0.0 (clang-1600.0.26.6)]
executable: [/Users/felipeangelim/Workspace/forecasting-course/.venv/bin/python](https://file+.vscode-resource.vscode-cdn.net/Users/felipeangelim/Workspace/forecasting-course/.venv/bin/python)
   machine: macOS-15.1-arm64-arm-64bit

Python dependencies:
          pip: 24.3.1
       sktime: 0.35.0
      sklearn: 1.5.2
       skbase: 0.12.0
        numpy: 1.26.4
        scipy: 1.14.1
       pandas: 2.2.3
   matplotlib: 3.10.0
       joblib: 1.4.2
        numba: None
  statsmodels: 0.14.4
     pmdarima: None
statsforecast: None
      tsfresh: None
      tslearn: None
        torch: 2.5.1
   tensorflow: None

</details>

### PR DESCRIPTION
**Pull request by RajdeepKushwaha5**
## Summary

Closes #7691.

Fixes a bug in `_NeuralForecastAdapter._get_id_idx` that caused a `KeyError`
(and potential silent data corruption) when predicting with hierarchical
(`pd_multiindex_hier`) time series data.

## Root Cause

The `_get_id_idx` method constructed a `unique_id` string for each time series
instance by concatenating all instance-level index values using numpy's
`.sum(axis=1)` on an object-dtype 2D array:

```python
# OLD (buggy)
id_idx = indices.to_frame().values
id_int = id_idx[:, :-1].sum(axis=1)
```

This had **two problems**:

### 1. Collision without separator
Tuples `("AB", "C")` and `("A", "BC")` both produce `"ABC"` — two
distinct series would share the same `unique_id`, corrupting the
mapping between neuralforecast's output and the sktime MultiIndex.

This is especially problematic with hierarchical data produced by
sktime's `Aggregator`, where `__total` rows appear alongside
regular level values at multiple positions.

### 2. Unreliable numpy object-array sum
`numpy.sum(axis=1)` on a 2D `object`-dtype array does not guarantee
Python-level string concatenation in all numpy versions and edge cases,
particularly for mixed-type index levels.

## Fix

Replaced the numpy sum with a pandas-based join using a null-byte separator
(`\x00`) that cannot appear in any normal index level value:

```python
# NEW (fixed)
id_frame = indices.to_frame(index=False)
instance_cols = id_frame.iloc[:, :-1]
id_int = instance_cols.astype(str).agg(self._ID_SEP.join, axis=1)
```

The separator constant `_ID_SEP = "\x00"` is defined as a class attribute
for clarity and easy reference.

## Changes

- `sktime/forecasting/base/adapters/_neuralforecast.py`:
  - Added `_ID_SEP = "\x00"` class attribute
  - Rewrote `_get_id_idx` to use pandas-based separator-aware string join

- `sktime/forecasting/tests/test_neuralforecast.py`:
  - Added regression test `test_neural_forecast_hierarchical_predict_no_keyerror`
    that exercises fit + predict on `pd_multiindex_hier` data (2-level hierarchy)
    with a `skipif` guard for the neuralforecast soft dependency

## How to verify (requires `neuralforecast` to be installed)

```python
from sktime.utils._testing.hierarchical import _make_hierarchical
from sktime.forecasting.neuralforecast import NeuralForecastLSTM

y = _make_hierarchical(hierarchy_levels=(2, 2), max_timepoints=20, min_timepoints=20, n_columns=1)
y.index = y.index.set_levels(range(20), level=-1)
y = y.sort_index()
y_train = y.loc[y.index.get_level_values(-1) < 16]

model = NeuralForecastLSTM(freq=1, max_steps=1, trainer_kwargs={"logger": False})
model.fit(y_train, fh=[1, 2, 3, 4])
y_pred = model.predict()  # Previously raised KeyError
print(y_pred)
```

## References

- Issue: #7691
- Related: #7776 (closed PR that partially addressed this), #7876

### CODE REVIEW

**Comment by marrov** [comment]
Closing this one as it looks AI generated and there is already a PR in progress for this issue.
